File: bindings/java/src/main/java/org/yorgl/YorGL.java
package org.yorgl;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.Locale;

public final class YorGL implements AutoCloseable {
    private long handle;

    private YorGL(long handle) {
        this.handle = handle;
    }

    public static void load() {
        load("yorgl");
    }

    public static void load(String libraryName) {
        try {
            System.loadLibrary(libraryName);
            return;
        } catch (UnsatisfiedLinkError original) {
            loadBundledNative(libraryName, original);
        }
    }

    public static YorGL create() {
        return create(BackendKind.Null);
    }

    public static YorGL create(BackendKind backend) {
        long ptr = YorGLNative.create(backend.id);
        if (ptr == 0L) {
            throw new IllegalStateException("YorGL failed to create " + backend.name() + " backend");
        }
        return new YorGL(ptr);
    }

    public boolean isValid() {
        return handle != 0L && YorGLNative.isValid(handle);
    }

    public String getBackendName() {
        return YorGLNative.backendName(handle);
    }

    public RendererCapabilities getCapabilities() {
        return new RendererCapabilities(YorGLNative.capabilities(handle));
    }

    public RenderDiagnostics getDiagnostics() {
        return new RenderDiagnostics(YorGLNative.diagnostics(handle));
    }

    public boolean createSwapChain(long windowHandle, int width, int height) {
        return YorGLNative.createSwapChain(handle, windowHandle, width, height);
    }

    public boolean createSwapChain(long windowHandle, SwapChainOptions options) {
        return YorGLNative.createSwapChainWithOptions(
            handle,
            windowHandle,
            options.width,
            options.height,
            options.bufferCount,
            options.presentMode.id,
            options.allowTearing);
    }

    public void resize(int width, int height) {
        YorGLNative.resize(handle, width, height);
    }

    public void beginFrame() {
        YorGLNative.beginFrame(handle);
    }

    public void setViewport(float x, float y, float width, float height) {
        YorGLNative.setViewport(handle, x, y, width, height);
    }

    public void clearColor(float r, float g, float b, float a) {
        YorGLNative.clearColor(handle, r, g, b, a);
    }

    public void clearDepth(float depth) {
        YorGLNative.clearDepth(handle, depth);
    }

    public void setPresentMode(PresentMode mode) {
        YorGLNative.setPresentMode(handle, mode.id);
    }

    public void endFrame() {
        YorGLNative.endFrame(handle);
    }

    public long createTexture(int width, int height, byte[] pixels) {
        return YorGLNative.createTexture(handle, width, height, pixels);
    }

    public boolean updateTextureRegion(long texture, int x, int y, int width, int height, byte[] pixels) {
        return YorGLNative.updateTextureRegion(handle, texture, x, y, width, height, pixels);
    }

    public void destroyTexture(long texture) {
        YorGLNative.destroyTexture(handle, texture);
    }

    public void guiBegin(int width, int height) {
        YorGLNative.guiBegin(handle, width, height);
    }

    public void guiDrawQuad(float x, float y, float w, float h, float u0, float v0, float u1, float v1, float r, float g, float b, float a) {
        YorGLNative.guiDrawQuad(handle, x, y, w, h, u0, v0, u1, v1, r, g, b, a);
    }

    public void guiDrawGradientQuad(float x, float y, float w, float h, float[] rgba16) {
        YorGLNative.guiDrawGradientQuad(handle, x, y, w, h, rgba16);
    }

    public void guiSetTexture(long texture) {
        YorGLNative.guiSetTexture(handle, texture);
    }

    public void guiSetScissor(float x, float y, float w, float h) {
        YorGLNative.guiSetScissor(handle, x, y, w, h);
    }

    public void guiClearScissor() {
        YorGLNative.guiClearScissor(handle);
    }

    public void guiSetSdfMode(boolean enabled) {
        YorGLNative.guiSetSdfMode(handle, enabled);
    }

    public void guiSetSdfParams(float edge, float softness, float weightBias) {
        YorGLNative.guiSetSdfParams(handle, edge, softness, weightBias);
    }

    public void guiBlurRect(float x, float y, float w, float h, int passes) {
        YorGLNative.guiBlurRect(handle, x, y, w, h, passes);
    }

    public void guiEnd() {
        YorGLNative.guiEnd(handle);
    }

    public void cubemapRender(long[] faces, float yawRadians, int width, int height) {
        YorGLNative.cubemapRender(handle, faces, yawRadians, width, height);
    }

    public void panoramaRender(long[] faces, float angle, int width, int height) {
        cubemapRender(faces, angle, width, height);
    }

    public void worldUploadMesh(float[] vertices, int floatCount) {
        YorGLNative.worldUploadMesh(handle, vertices, floatCount);
    }

    public void worldUploadSection(long sectionId, int x, int y, int z, float[] vertices, int floatCount) {
        YorGLNative.worldUploadSection(handle, sectionId, x, y, z, vertices, floatCount);
    }

    public void worldUploadSectionLayer(long sectionId, int x, int y, int z, int layer, float[] vertices, int floatCount) {
        YorGLNative.worldUploadSectionLayer(handle, sectionId, x, y, z, layer, vertices, floatCount);
    }

    public void worldUploadSectionLayerTextured(long sectionId, int x, int y, int z, int layer, long texture, float[] vertices, int floatCount) {
        YorGLNative.worldUploadSectionLayerTextured(handle, sectionId, x, y, z, layer, texture, vertices, floatCount);
    }

    public void worldRemoveSection(long sectionId) {
        YorGLNative.worldRemoveSection(handle, sectionId);
    }

    public void worldClearSections() {
        YorGLNative.worldClearSections(handle);
    }

    public void worldSetTexture(long texture) {
        YorGLNative.worldSetTexture(handle, texture);
    }

    public void worldSetTextureFilter(TextureFilter filter) {
        YorGLNative.worldSetTextureFilter(handle, filter.id);
    }

    public void worldSetSkyColor(float r, float g, float b) {
        YorGLNative.worldSetSkyColor(handle, r, g, b);
    }

    public void worldRender(float cameraX, float cameraY, float cameraZ, float dirX, float dirY, float dirZ, float fovYDegrees, float farPlane, int width, int height) {
        YorGLNative.worldRender(handle, cameraX, cameraY, cameraZ, dirX, dirY, dirZ, fovYDegrees, farPlane, width, height);
    }

    public long sdfFontCreate(byte[] ttfData, float fontSize) {
        return YorGLNative.sdfFontCreate(handle, ttfData, fontSize);
    }

    public void sdfFontDestroy(long font) {
        YorGLNative.sdfFontDestroy(handle, font);
    }

    public long sdfFontAtlas(long font) {
        return YorGLNative.sdfFontAtlas(handle, font);
    }

    public float[] sdfFontMetrics(long font) {
        return YorGLNative.sdfFontMetrics(handle, font);
    }

    public float[] sdfFontGlyph(long font, int codepoint) {
        return YorGLNative.sdfFontGlyph(handle, font, codepoint);
    }

    public float sdfFontKerning(long font, int leftCodepoint, int rightCodepoint) {
        return YorGLNative.sdfFontKerning(handle, font, leftCodepoint, rightCodepoint);
    }

    public float sdfFontTextWidth(long font, String text, float scale) {
        return YorGLNative.sdfFontTextWidth(handle, font, text, scale);
    }

    public float sdfFontLineHeight(long font, float scale) {
        return YorGLNative.sdfFontLineHeight(handle, font, scale);
    }

    public void sdfFontDrawText(long font, String text, float x, float y, float scale, float r, float g, float b, float a, float weight, boolean shadow) {
        YorGLNative.sdfFontDrawText(handle, font, text, x, y, scale, r, g, b, a, weight, shadow);
    }

    @Override
    public void close() {
        if (handle != 0L) {
            YorGLNative.destroy(handle);
            handle = 0L;
        }
    }

    private static void loadBundledNative(String libraryName, UnsatisfiedLinkError original) {
        String resource = nativeResourceName(libraryName);
        try (InputStream input = YorGL.class.getResourceAsStream(resource)) {
            if (input == null) {
                throw original;
            }
            String suffix = resource.substring(resource.lastIndexOf('.'));
            Path file = Files.createTempFile("yorgl-", suffix);
            file.toFile().deleteOnExit();
            Files.copy(input, file, StandardCopyOption.REPLACE_EXISTING);
            System.load(file.toAbsolutePath().toString());
        } catch (IOException e) {
            original.addSuppressed(e);
            throw original;
        }
    }

    private static String nativeResourceName(String libraryName) {
        String os = System.getProperty("os.name").toLowerCase(Locale.ROOT);
        String arch = System.getProperty("os.arch").toLowerCase(Locale.ROOT);
        String platform;
        if (os.contains("win") && (arch.contains("64") || arch.contains("amd64"))) {
            platform = "windows-x64";
        } else {
            throw new IllegalStateException("No bundled YorGL native for os=" + os + " arch=" + arch);
        }
        String file = os.contains("win") ? libraryName + ".dll" : "lib" + libraryName + ".so";
        return "/org/yorgl/native/" + platform + "/" + file;
    }
}

File: bindings/kotlin/src/main/kotlin/org/yorgl/YorGL.kt
package org.yorgl

import java.nio.file.Files

enum class BackendKind(internal val id: Int) {
    Null(0),
    Dx11(1),
}

class YorGL private constructor(private var handle: Long) : AutoCloseable {
    val valid: Boolean get() = handle != 0L && YorGLNative.isValid(handle)
    val backendName: String get() = YorGLNative.backendName(handle)

    fun createSwapChain(windowHandle: Long, width: Int, height: Int): Boolean =
        YorGLNative.createSwapChain(handle, windowHandle, width, height)

    fun resize(width: Int, height: Int) = YorGLNative.resize(handle, width, height)
    fun beginFrame() = YorGLNative.beginFrame(handle)
    fun clearColor(r: Float, g: Float, b: Float, a: Float) = YorGLNative.clearColor(handle, r, g, b, a)
    fun endFrame() = YorGLNative.endFrame(handle)
    fun createTexture(width: Int, height: Int, pixels: ByteArray): Long =
        YorGLNative.createTexture(handle, width, height, pixels)
    fun destroyTexture(texture: Long) = YorGLNative.destroyTexture(handle, texture)

    fun guiBegin(width: Int, height: Int) = YorGLNative.guiBegin(handle, width, height)
    fun guiDrawQuad(x: Float, y: Float, w: Float, h: Float, u0: Float, v0: Float, u1: Float, v1: Float, r: Float, g: Float, b: Float, a: Float) =
        YorGLNative.guiDrawQuad(handle, x, y, w, h, u0, v0, u1, v1, r, g, b, a)
    fun guiDrawGradientQuad(x: Float, y: Float, w: Float, h: Float, rgba16: FloatArray) =
        YorGLNative.guiDrawGradientQuad(handle, x, y, w, h, rgba16)
    fun guiSetTexture(texture: Long) = YorGLNative.guiSetTexture(handle, texture)
    fun guiSetScissor(x: Float, y: Float, w: Float, h: Float) = YorGLNative.guiSetScissor(handle, x, y, w, h)
    fun guiClearScissor() = YorGLNative.guiClearScissor(handle)
    fun guiSetSdfMode(enabled: Boolean) = YorGLNative.guiSetSdfMode(handle, enabled)
    fun guiSetSdfParams(edge: Float, softness: Float, weightBias: Float) =
        YorGLNative.guiSetSdfParams(handle, edge, softness, weightBias)
    fun guiBlurRect(x: Float, y: Float, w: Float, h: Float, passes: Int) =
        YorGLNative.guiBlurRect(handle, x, y, w, h, passes)
    fun guiEnd() = YorGLNative.guiEnd(handle)

    fun panoramaRender(faces: LongArray, angle: Float, width: Int, height: Int) =
        YorGLNative.panoramaRender(handle, faces, angle, width, height)

    fun worldUploadMesh(vertices: FloatArray, floatCount: Int) =
        YorGLNative.worldUploadMesh(handle, vertices, floatCount)
    fun worldUploadSection(sectionId: Long, x: Int, y: Int, z: Int, vertices: FloatArray, floatCount: Int) =
        YorGLNative.worldUploadSection(handle, sectionId, x, y, z, vertices, floatCount)
    fun worldUploadSectionLayer(sectionId: Long, x: Int, y: Int, z: Int, layer: Int, vertices: FloatArray, floatCount: Int) =
        YorGLNative.worldUploadSectionLayer(handle, sectionId, x, y, z, layer, vertices, floatCount)
    fun worldRemoveSection(sectionId: Long) = YorGLNative.worldRemoveSection(handle, sectionId)
    fun worldClearSections() = YorGLNative.worldClearSections(handle)
    fun worldSetTexture(texture: Long) = YorGLNative.worldSetTexture(handle, texture)
    fun worldSetSkyColor(r: Float, g: Float, b: Float) = YorGLNative.worldSetSkyColor(handle, r, g, b)
    fun worldRender(cameraX: Float, cameraY: Float, cameraZ: Float, dirX: Float, dirY: Float, dirZ: Float, fovYDegrees: Float, farPlane: Float, width: Int, height: Int) =
        YorGLNative.worldRender(handle, cameraX, cameraY, cameraZ, dirX, dirY, dirZ, fovYDegrees, farPlane, width, height)

    fun sdfFontCreate(ttfData: ByteArray, fontSize: Float): Long = YorGLNative.sdfFontCreate(handle, ttfData, fontSize)
    fun sdfFontDestroy(font: Long) = YorGLNative.sdfFontDestroy(handle, font)
    fun sdfFontAtlas(font: Long): Long = YorGLNative.sdfFontAtlas(handle, font)
    fun sdfFontMetrics(font: Long): FloatArray? = YorGLNative.sdfFontMetrics(handle, font)
    fun sdfFontGlyph(font: Long, codepoint: Int): FloatArray? = YorGLNative.sdfFontGlyph(handle, font, codepoint)
    fun sdfFontKerning(font: Long, leftCodepoint: Int, rightCodepoint: Int): Float =
        YorGLNative.sdfFontKerning(handle, font, leftCodepoint, rightCodepoint)

    override fun close() {
        if (handle != 0L) {
            YorGLNative.destroy(handle)
            handle = 0
        }
    }

    companion object {
        fun load(libraryName: String = "yorgl") {
            runCatching { System.loadLibrary(libraryName) }.getOrElse { original ->
                val resource = nativeResourceName(libraryName)
                val stream = YorGL::class.java.getResourceAsStream(resource) ?: throw original
                val suffix = resource.substringAfterLast('.', ".dll")
                val file = Files.createTempFile("yorgl-", suffix)
                file.toFile().deleteOnExit()
                stream.use { input -> Files.copy(input, file, java.nio.file.StandardCopyOption.REPLACE_EXISTING) }
                System.load(file.toAbsolutePath().toString())
            }
        }

        fun create(backend: BackendKind = BackendKind.Null): YorGL {
            val ptr = YorGLNative.create(backend.id)
            check(ptr != 0L) { "YorGL failed to create ${backend.name} backend" }
            return YorGL(ptr)
        }

        private fun nativeResourceName(libraryName: String): String {
            val os = System.getProperty("os.name").lowercase()
            val arch = System.getProperty("os.arch").lowercase()
            val platform = when {
                os.contains("win") && (arch.contains("64") || arch.contains("amd64")) -> "windows-x64"
                else -> error("No bundled YorGL native for os=$os arch=$arch")
            }
            val file = if (os.contains("win")) "$libraryName.dll" else "lib$libraryName.so"
            return "/org/yorgl/native/$platform/$file"
        }
    }
}
